src/algebra/direct_limit.lean

```diff
@@ -0,0 +1,483 @@
+    let ⟨k, hik, hjk⟩ := directed_order.directed i j in
+    ⟨k, λ l hl,
+      (finset.mem_union.1 (dfinsupp.support_add hl)).elim
+        (λ hl, le_trans (hi _ hl) hik)
+        (λ hl, le_trans (hj _ hl) hjk),
+      by simp [linear_map.map_add, hxi, hyj,
+          to_module_totalize_of_le hik hi,
+          to_module_totalize_of_le hjk hj]⟩)
+  (λ a x ⟨i, hi, hxi⟩,
+    ⟨i, λ k hk, hi k (dfinsupp.support_smul hk),
+      by simp [linear_map.map_smul, hxi]⟩)
+
+theorem of.zero_exact {i x} (H : of R ι G f i x = 0) :
+  ∃ j hij, f i j hij x = (0 : G j) :=
+let ⟨j, hj, hxj⟩ := of.zero_exact_aux H in
+if hx0 : x = 0 then ⟨i, le_refl _, by simp [hx0]⟩
+else
+  have hij : i ≤ j, from hj _ $
+    by simp [direct_sum.apply_eq_component, hx0],
+  ⟨j, hij, by simpa [totalize_apply, hij] using hxj⟩
+
+end direct_limit
+
+end module
+
+
+namespace add_comm_group
+
+variables [Π i, add_comm_group (G i)]
+
+def direct_limit (f : Π i j, i ≤ j → G i → G j)
+  [Π i j hij, is_add_group_hom (f i j hij)] [directed_system G f] : Type* :=
+@module.direct_limit ℤ _ ι _ _ _ G _ _ _
+  (λ i j hij, is_add_group_hom.to_linear_map $ f i j hij)
+  ⟨directed_system.Hid f, directed_system.Hcomp f⟩
+
+namespace direct_limit
+
+variables (f : Π i j, i ≤ j → G i → G j)
+variables [Π i j hij, is_add_group_hom (f i j hij)] [directed_system G f]
+
+def directed_system : module.directed_system G (λ i j hij, is_add_group_hom.to_linear_map $ f i j hij) :=
+⟨directed_system.Hid f, directed_system.Hcomp f⟩
+
+local attribute [instance] directed_system
+
+instance : add_comm_group (direct_limit G f) :=
+module.direct_limit.add_comm_group G (λ i j hij, is_add_group_hom.to_linear_map $ f i j hij)
+
+set_option class.instance_max_depth 50
+def of (i) : G i → direct_limit G f :=
+module.direct_limit.of ℤ ι G (λ i j hij, is_add_group_hom.to_linear_map $ f i j hij) i
+variables {G f}
+
+instance of.is_add_group_hom (i) : is_add_group_hom (of G f i) :=
+linear_map.is_add_group_hom _
+
+@[simp] lemma of_f {i j} (hij) (x) : of G f j (f i j hij x) = of G f i x :=
+module.direct_limit.of_f
+
+@[simp] lemma of_zero (i) : of G f i 0 = 0 := is_add_group_hom.zero _
+@[simp] lemma of_add (i x y) : of G f i (x + y) = of G f i x + of G f i y := is_add_group_hom.add _ _ _
+@[simp] lemma of_neg (i x) : of G f i (-x) = -of G f i x := is_add_group_hom.neg _ _
+@[simp] lemma of_sub (i x y) : of G f i (x - y) = of G f i x - of G f i y := is_add_group_hom.sub _ _ _
+
+@[elab_as_eliminator]
+protected theorem induction_on {C : direct_limit G f → Prop} (z : direct_limit G f)
+  (ih : ∀ i x, C (of G f i x)) : C z :=
+module.direct_limit.induction_on z ih
+
+theorem of.zero_exact (i x) (h : of G f i x = 0) : ∃ j hij, f i j hij x = 0 :=
+module.direct_limit.of.zero_exact h
+
+variables (P : Type u₁) [add_comm_group P]
+variables (g : Π i, G i → P) [Π i, is_add_group_hom (g i)]
+variables (Hg : ∀ i j hij x, g j (f i j hij x) = g i x)
+
+variables (G f)
+def rec : direct_limit G f → P :=
```

Comment:
This should be called `lift` probably, and another new line

Reply:
renamed to `rec`